File: examples/mqtt_debounced_button/app/spiffsconfig.cpp
#include <SmingCore/SmingCore.h>
#include "spiffsconfig.h"

SpiffsConfigStorage SpiffsConfig;


const String DEFAULTLIGHT_SETTINGS_FILE = "defaultlight.conf";
const String NET_SETTINGS_FILE = "net.conf";
const String WIFISSID_SETTINGS_FILE = "wifi.ssid.conf";
const String WIFIPASS_SETTINGS_FILE = "wifi.pass.conf";
const String MQTTCLIENT_SETTINGS_FILE = "mqtt.clientid.conf";
const String MQTTUSER_SETTINGS_FILE = "mqtt.user.conf";
const String MQTTPASS_SETTINGS_FILE = "mqtt.pass.conf";
const String MQTTBROKER_SETTINGS_FILE = "mqttbroker.conf";
const String AUTHTOKEN_SETTINGS_FILE = "authtoken.conf";
const String UPDATE_INTERVAL_SETTINGS_FILE = "updateinterval.conf";
const String DEBOUNCE_INTERVAL_SETTINGS_FILE = "debounceinterval.conf";
const String USEDHCP_SETTINGS_FILE = "dhcp.flag";



void SpiffsConfigStorage::load()
{
	if (exist())
	{
		uint32_t netsettings[4];
		file_t f = fileOpen(NET_SETTINGS_FILE, eFO_ReadOnly);
		fileRead(f, (void*) netsettings, 4*sizeof(uint32_t));
		fileClose(f);
		ip = IPAddress(netsettings[0]);
		netmask = IPAddress(netsettings[1]);
		gw = IPAddress(netsettings[2]);
		mqtt_port = (uint16_t)(netsettings[3]);
		wifi_ssid = fileGetContent(WIFISSID_SETTINGS_FILE);
		wifi_pass = fileGetContent(WIFIPASS_SETTINGS_FILE);
		mqtt_broker = fileGetContent(MQTTBROKER_SETTINGS_FILE);
		mqtt_clientid = fileGetContent(MQTTCLIENT_SETTINGS_FILE);
		mqtt_user = fileGetContent(MQTTUSER_SETTINGS_FILE);
		mqtt_pass = fileGetContent(MQTTPASS_SETTINGS_FILE);
		authtoken = fileGetContent(AUTHTOKEN_SETTINGS_FILE);
		enabledhcp = fileExist(USEDHCP_SETTINGS_FILE);
		f = fileOpen(UPDATE_INTERVAL_SETTINGS_FILE, eFO_ReadOnly);
		fileRead(f, (void*) &publish_interval, sizeof(uint32_t));
		fileClose(f);
		f = fileOpen(DEBOUNCE_INTERVAL_SETTINGS_FILE, eFO_ReadOnly);
		fileRead(f, (void*) &debounce_interval, sizeof(uint32_t));
		fileClose(f);
	}
}

void SpiffsConfigStorage::save()
{
	uint32_t netsettings[4] = {ip,netmask,gw, (uint32_t)mqtt_port};
	file_t f = fileOpen(NET_SETTINGS_FILE, eFO_WriteOnly | eFO_CreateNewAlways);
	fileWrite(f, (void*) netsettings, 4*sizeof(uint32_t));
	fileClose(f);
	fileSetContent(WIFISSID_SETTINGS_FILE, wifi_ssid);
	fileSetContent(WIFIPASS_SETTINGS_FILE, wifi_pass);
	fileSetContent(MQTTBROKER_SETTINGS_FILE, mqtt_broker);
	fileSetContent(MQTTCLIENT_SETTINGS_FILE, mqtt_clientid);
	fileSetContent(MQTTUSER_SETTINGS_FILE, mqtt_user);
	fileSetContent(MQTTPASS_SETTINGS_FILE, mqtt_pass);
	fileSetContent(AUTHTOKEN_SETTINGS_FILE, authtoken);
	if (enabledhcp)
		fileSetContent(USEDHCP_SETTINGS_FILE, "true");
	else
		fileDelete(USEDHCP_SETTINGS_FILE);
	f = fileOpen(UPDATE_INTERVAL_SETTINGS_FILE, eFO_WriteOnly | eFO_CreateNewAlways);
	fileWrite(f, (void*) &publish_interval, sizeof(uint32_t));
	fileClose(f);
	f = fileOpen(DEBOUNCE_INTERVAL_SETTINGS_FILE, eFO_WriteOnly | eFO_CreateNewAlways);
	fileWrite(f, (void*) &debounce_interval, sizeof(uint32_t));
	fileClose(f);
}

bool SpiffsConfigStorage::exist() { return fileExist(NET_SETTINGS_FILE); }

File: examples/mqtt_debounced_button/spiffsconfig.py
#!/usr/bin/python3

from ipaddress import ip_address
import os
import struct

def writeConfig(ip,nm,gw,wifi_ssid,wifi_pass,mqtt_broker,mqtt_clientid,mqtt_user,mqtt_pass,authtoken,dhcp=True,mqtt_port=1883,update_interval=7000,debounce_interval=3500):
    NET_SETTINGS_FILE = "net.conf"
    WIFISSID_SETTINGS_FILE = "wifi.ssid.conf"
    WIFIPASS_SETTINGS_FILE = "wifi.pass.conf"
    MQTTCLIENT_SETTINGS_FILE = "mqtt.clientid.conf"
    MQTTUSER_SETTINGS_FILE = "mqtt.user.conf"
    MQTTPASS_SETTINGS_FILE = "mqtt.pass.conf"
    MQTTBROKER_SETTINGS_FILE = "mqttbroker.conf"
    AUTHTOKEN_SETTINGS_FILE = "authtoken.conf"
    USEDHCP_SETTINGS_FILE = "dhcp.flag"
    UPDATE_INTERVAL_SETTINGS_FILE = "updateinterval.conf"
    DEBOUNCE_INTERVAL_SETTINGS_FILE = "debounceinterval.conf"
    DIR="./files/"
    with open(os.path.join(DIR, NET_SETTINGS_FILE),"wb") as fh:
        fh.write(struct.pack(">III", int(ip_address(ip)), int(ip_address(nm)), int(ip_address(gw))))
        fh.write(struct.pack("<I",  int(mqtt_port)))
    with open(os.path.join(DIR, UPDATE_INTERVAL_SETTINGS_FILE),"wb") as fh:
        fh.write(struct.pack("<I",  int(update_interval)))
    with open(os.path.join(DIR, DEBOUNCE_INTERVAL_SETTINGS_FILE),"wb") as fh:
        fh.write(struct.pack("<I",  int(debounce_interval)))
    with open(os.path.join(DIR, WIFISSID_SETTINGS_FILE),"wb") as fh:
        fh.write(wifi_ssid)
    with open(os.path.join(DIR, WIFIPASS_SETTINGS_FILE),"wb") as fh:
        fh.write(wifi_pass)
    with open(os.path.join(DIR, MQTTBROKER_SETTINGS_FILE),"wb") as fh:
        fh.write(mqtt_broker)
    with open(os.path.join(DIR, MQTTCLIENT_SETTINGS_FILE),"wb") as fh:
        fh.write(mqtt_clientid)
    with open(os.path.join(DIR, MQTTUSER_SETTINGS_FILE),"wb") as fh:
        fh.write(mqtt_user)
    with open(os.path.join(DIR, MQTTPASS_SETTINGS_FILE),"wb") as fh:
        fh.write(mqtt_pass)
    with open(os.path.join(DIR, AUTHTOKEN_SETTINGS_FILE),"wb") as fh:
        fh.write(authtoken)
    if dhcp:
        with open(os.path.join(DIR, USEDHCP_SETTINGS_FILE),"wb") as fh:
            fh.write(b"true")
    else:
        try:
            os.unlink(os.path.join(DIR, USEDHCP_SETTINGS_FILE))
        except:
            pass

## Example Use
# writeConfig(
#     ip="",
#     nm="",
#     gw="",
#     dhcp=True,
#     wifi_ssid=b"",
#     wifi_pass=b"",
#     mqtt_port=1883,
#     mqtt_broker=b"",
#     mqtt_clientid=b"",
#     mqtt_user=b"",
#     mqtt_pass=b"",
#     authtoken=b"")
